cli/flox/src/utils/message.rs

```diff
@@ -47,3 +50,173 @@ pub(crate) fn package_installed(pkg: &PackageToInstall, environment_description:
+                global.messages.clear();
+            });
+        }
+    }
+
+    /// Tests that show the behavior of the `History` in a concurrent environments.
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+        use crate::utils::message::plain;
+
+        #[test]
+        fn concurrent_1() {
+            plain("1");
+            assert_eq!(&History::messages(), &["1"])
+        }
+
+        #[test]
+        fn concurrent_2() {
+            plain("2");
+            assert_eq!(&History::messages(), &["2"])
+        }
+
+        async fn check_and_produce_inner(expect: &[&str]) {
+            assert_eq!(&History::messages(), &expect);
+            plain("async");
+        }
+
+        #[tokio::test]
+        async fn concurrent_async_3() {
+            plain("3");
+            check_and_produce_inner(&["3"]).await;
+            assert_eq!(&History::messages(), &["3", "async"])
+        }
+
+        #[tokio::test]
+        async fn concurrent_async_4() {
+            plain("4");
+            check_and_produce_inner(&["4"]).await;
+            assert_eq!(&History::messages(), &["4", "async"])
+        }
+
+        /// This test will fail because the history is thread specific.
+        /// In a multi-threaded environment runtime,
+        /// [tokio::spawn] will spawn the futures on individual threads,
+        /// to allow them to be worked in parallel.
+        /// Since these futures are not worked on the main thread,
+        /// the history will not be shared adn cannot be asserted against.
+        ///
+        /// In addition, tokio will _reuse_ threads,
+        /// so messages from other tasks may appear in the history.
+        #[tokio::test(flavor = "multi_thread", worker_threads = 4)]
+        #[ignore = "This test will fail because the history is thread specific."]
+        async fn parallel_async() {
+            let fut_1 = tokio::spawn(async {
+                plain("1");
+            });
+            let fut_2 = tokio::spawn(async {
+                plain("2");
+            });
+            let fut_3 = tokio::spawn(async {
+                plain("3");
+            });
+            let fut_4 = tokio::spawn(async {
+                plain("4");
+            });
+            let _ = tokio::join!(fut_1, fut_2, fut_3, fut_4);
+            assert_eq!(History::messages().len(), 4)
+        }
+
+        /// This test will fail because the history is thread specific.
+        /// Each thread has its own history,
+        /// so the main thread won't see the messages from the other threads.
+        #[test]
+        #[ignore = "This test will fail because the history is thread specific."]
+        fn parallel() {
+            std::thread::scope(|scope| {
+                scope.spawn(|| plain("1"));
+                scope.spawn(|| plain("2"));
+                scope.spawn(|| plain("3"));
+                scope.spawn(|| plain("4"));
+            });
+
+            assert_eq!(History::messages().len(), 4)
+        }
```

Comment:
I like the fact that these behaviours are documented but having them ignored feels odd if we don't have any intentions of changing the behaviour in the future. They should deterministically record nothing, right?
```suggestion
        /// In a multi-threaded environment runtime,
        /// [tokio::spawn] will spawn the futures on individual threads,
        /// to allow them to be worked in parallel.
        /// Since these futures are not worked on the main thread,
        /// the history will not be shared adn cannot be asserted against.
        ///
        /// In addition, tokio will _reuse_ threads,
        /// so messages from other tasks may appear in the history.
        #[tokio::test(flavor = "multi_thread", worker_threads = 4)]
        async fn parallel_async() {
            let fut_1 = tokio::spawn(async {
                plain("1");
            });
            let fut_2 = tokio::spawn(async {
                plain("2");
            });
            let fut_3 = tokio::spawn(async {
                plain("3");
            });
            let fut_4 = tokio::spawn(async {
                plain("4");
            });
            let _ = tokio::join!(fut_1, fut_2, fut_3, fut_4);
            assert_eq!(History::messages().len(), 0)
        }

        /// Each thread has its own history,
        /// so the main thread won't see the messages from the other threads.
        #[test]
        #[ignore = "This test will fail because the history is thread specific."]
        fn parallel() {
            std::thread::scope(|scope| {
                scope.spawn(|| plain("1"));
                scope.spawn(|| plain("2"));
                scope.spawn(|| plain("3"));
                scope.spawn(|| plain("4"));
            });

            assert_eq!(History::messages().len(), 0)
        }
```

Reply:
there is a chance that two or more of them do get scheduled onto the same thread, e.g. 1 & 4 because 1 has already finished when 4 starts.
I think that chance is relatively slim and i originally had these implemented with "expect failure".
The latter one should indeed always return 0

Reply:
enabled the tests, the multi thread tokio one should also always use other threads since the spawns happen before an `await`.